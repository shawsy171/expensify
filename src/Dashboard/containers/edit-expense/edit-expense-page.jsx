import React from 'react';

const EditExpensePage = () => (
  <p> this is the edit page </p>
);

export default EditExpensePage;
