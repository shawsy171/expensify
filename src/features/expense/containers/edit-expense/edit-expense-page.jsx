import React from 'react';
import ReactRouterPropTypes from 'react-router-prop-types';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

// store
import { editExpense, removeExpense } from './../../store/expense/actions';

// components
import ExpenseForm from './../../components/expense-form/expense-form';

export class EditExpensePage extends React.Component {
  onSubmit = (submittedExpense) => {
    this.props.editExpense(this.props.match.params.id, submittedExpense);
    this.props.history.push('/');
  };

  removeClick = () => {
    this.props.removeExpense(this.props.match.params.id);
    this.props.history.push('/');
  }

  render() {
    const {
      expense,
      match: {
        params: {
          id,
        },
      },
    } = this.props;
    return (
      <div>
        <p> this is the edit page, {id}</p>
        <ExpenseForm
          expense={expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.removeClick} >Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  editExpense: (id, submittedExpense) => dispatch(editExpense(id, submittedExpense)),
  removeExpense: id => dispatch(removeExpense(id)),
});

EditExpensePage.propTypes = {
  expense: PropTypes.shape({
    amount: PropTypes.number,
    createdAt: PropTypes.number,
    description: PropTypes.string,
    id: PropTypes.string,
    note: PropTypes.string,
  }).isRequired,
  editExpense: PropTypes.func.isRequired,
  removeExpense: PropTypes.func.isRequired,
  match: ReactRouterPropTypes.match.isRequired,
  history: ReactRouterPropTypes.history.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
