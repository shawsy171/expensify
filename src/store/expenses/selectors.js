/** filters and sorts an array of expenses 
 * @param { Array<Object> } expenses 
 * @interface Array<{
 *  text: string, 
 *  sortBy: string,
 *  startDate: number,
 *  endDate: number,
 * }>
 * @example [{
 *  description = 'Rent', 
 *  note = 'to be paid',
 *  amount = 5000,
 *  createdAt = 19299303
 * }, {
 *  description = 'Coffee', 
 *  note = 'I love coffee in the morning',
 *  amount = 300,
 *  createdAt = 85849307
 * }]
 * @param { Object } filterObject
 * @interface {
 *  text?: string, 
 *  sortBy: string,
 *  startDate?: number,
 *  endDate?: number,
 * }
 * @example {
 *  text: 'rent', 
 *  sortBy: 'amount',
 *  startDate: 100,
 *  endDate: 1250,
 * }
 * @returns { Array<Object> } expenses
 */
export const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
}