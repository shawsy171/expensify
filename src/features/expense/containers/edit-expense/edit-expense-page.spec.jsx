import React from 'react';
import { shallow } from 'enzyme';
import { historyMock, matchMock } from './../../../../tests/router.mock';
import expenses from './../../../../tests/fixtures';

// component
import { EditExpensePage } from './edit-expense-page';

describe('EditExpensePage Component', () => {
  let wrapper;
  let history;
  let editExpense;
  let match;
  let removeExpense;

  beforeEach(() => {
    editExpense = jest.fn();
    history = { ...historyMock };
    match = { ...matchMock, params: { id: expenses[0].id } };
    removeExpense = jest.fn();

    wrapper = shallow(<EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      match={match}
      expense={expenses[0]}
    />);
  });

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should handle onSubmit editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
  });

  test('should remove expense ', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith(expenses[0].id);
  });
});
