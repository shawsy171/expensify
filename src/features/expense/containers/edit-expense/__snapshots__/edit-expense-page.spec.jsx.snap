// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage Component should render correctly 1`] = `
<div>
  <p>
     this is the edit page, 
    1
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "gym",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
